patient-care/webapp/client/shared/widgets.js

```diff
@@ -351,14 +400,115 @@ Template.listSamplesButton.helpers({
       action: "nothing"
     };
   },
+  alwaysShowAll() {
+    return this.sampleLabels && this.sampleLabels.length <= 6;
+  },
+  not(variable) {
+    return !variable;
+  },
+  tooManyToShowAll() {
+    return this.sampleLabels.length > 1000;
+  },
 });
 
 Template.listSamplesButton.events({
-  "click .show-list"(event, instance) {
-    instance.showAllSamples.set(!instance.showAllSamples.get());
+  "click .toggle-list"(event, instance) {
+    instance.showMore.set(!instance.showMore.get());
   },
   "click .toggle-study-labels"(event, instance) {
-    instance.hideStudyLabels.set(!instance.hideStudyLabels.get());
+    let { profile } = Meteor.user();
+    let newValue = !profile || !profile.showStudyLabels;
+
+    Meteor.users.update(Meteor.userId(), {
+      $set: {
+        "profile.showStudyLabels": newValue
+      }
+    });
+  },
+  "click .download-list"(event, instance) {
+    let { sampleLabels } = instance.data;
+
+    // unqualify sample labels before downloading the list
+    let { profile } = Meteor.user();
+    if (!profile || !profile.showStudyLabels) {
+      sampleLabels = MedBook.utility.unqualifySampleLabels(sampleLabels);
+    }
+
+    saveStringAsFile(sampleLabels.join("\n"), instance.data.filename);
+  },
+});
+
+// Template.listFeaturesButton
+
+let saveStringAsFile = function () {
```

Comment:
I'm not 100% sure what to do with this function. Should we attach it to the `MedBook` namespace? For now we only use it here, so here's a good place.